src/NuGet.Services.AzureSearch/DocumentUtilities.cs

```diff
@@ -0,0 +1,216 @@
+            document.FileSize = leaf.PackageSize;
+            document.FlattenedDependencies = GetFlattenedDependencies(leaf);
+            document.Hash = leaf.PackageHash;
+            document.HashAlgorithm = leaf.PackageHashAlgorithm;
+            document.IconUrl = leaf.IconUrl;
+            document.Language = leaf.Language;
+            document.LastEdited = leaf.LastEdited;
+            document.LicenseUrl = leaf.LicenseUrl;
+            document.MinClientVersion = leaf.MinClientVersion;
+            document.NormalizedVersion = normalizedVersion;
+            document.OriginalVersion = leaf.VerbatimVersion ?? leaf.PackageVersion;
+            document.PackageId = leaf.PackageId;
+            document.Prerelease = leaf.IsPrerelease;
+            document.ProjectUrl = leaf.ProjectUrl;
+            document.Published = leaf.Published;
+            document.ReleaseNotes = leaf.ReleaseNotes;
+            document.RequiresLicenseAcceptance = leaf.RequireLicenseAgreement;
+            document.SemVerLevel = leaf.IsSemVer2() ? 2 : (int?)null;
+            document.Summary = leaf.Summary;
+            document.Tags = leaf.Tags == null ? null : leaf.Tags.ToArray();
+            document.Title = leaf.Title;
+        }
+
+        public static string GetSearchDocumentKey(string packageId, SearchFilters searchFilters)
+        {
+            var lowerId = packageId.ToLowerInvariant();
+            var encodedId = EncodeKey(lowerId);
+            return $"{encodedId}-{searchFilters}";
+        }
+
+        public static string GetHijackDocumentKey(string packageId, string normalizedVersion)
+        {
+            var lowerId = packageId.ToLowerInvariant();
+            var lowerVersion = normalizedVersion.ToLowerInvariant();
+            return EncodeKey($"{lowerId}/{lowerVersion}");
+        }
+
+        private static string EncodeKey(string rawKey)
+        {
+            // First, encode the raw value for uniqueness.
+            var bytes = Encoding.UTF8.GetBytes(rawKey);
+            var unique = HttpServerUtility.UrlTokenEncode(bytes);
+
+            // Then, prepend a string as close to the raw key as possible, for readibility.
+            var readable = ReplaceUnsafeKeyCharacters(rawKey).TrimStart('_');
+
+            return readable.Length > 0 ? $"{readable}-{unique}" : unique;
+        }
+
+        private static string ReplaceUnsafeKeyCharacters(string input)
+        {
+            return Regex.Replace(
+                input,
+                "[^A-Za-z0-9-_]", // Remove equal sign as well, since it's ugly.
+                "_",
+                RegexOptions.None,
+                TimeSpan.FromSeconds(30));
+        }
+
+        private static DateTimeOffset? AssumeUtc(DateTime? dateTime)
+        {
+            if (!dateTime.HasValue)
+            {
+                return null;
+            }
+
+            return new DateTimeOffset(
+                DateTime.SpecifyKind(dateTime.Value, DateTimeKind.Utc),
```

Comment:
The call to `SpecifyKind(…)` is superfluous.  All it does is create a new `DateTime` object with the same `Ticks` value and a new `DateTimeKind` that you specify.  But the time zone is specified again below.

Just use [`DateTimeOffset(long, TimeSpan)`](https://docs.microsoft.com/en-us/dotnet/api/system.datetimeoffset.-ctor?view=netframework-4.7.2#System_DateTimeOffset__ctor_System_Int64_System_TimeSpan_):

    return new DateTimeOffset(dateTime.Ticks, TimeSpan.Zero);

Reply:
Great point!